### PR TITLE
DOCS: CLUSTER command supported for append-optimized tables

### PR DESCRIPTION
**Pull request by VMWareJennyBeth**

### CODE REVIEW

File: gpdb-doc/markdown/admin_guide/ddl/ddl-storage.html.md [hunk 1]
```diff
@@ -44,7 +44,7 @@ Use the `WITH` clause of the `CREATE TABLE` command to declare the table storage
 
 > **Note** You use the `appendoptimized=value` syntax to specify the append-optimized table storage type. `appendoptimized` is a thin alias for the `appendonly` legacy storage option. Greenplum Database stores `appendonly` in the catalog, and displays the same when listing storage options for append-optimized tables.
 
-`UPDATE` and `DELETE` are not allowed on append-optimized tables in a repeatable read or serizalizable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE` and triggers are not supported with append-optimized tables. `CLUSTER` on append-optimized tables is only supported over B-tree indexes.
+`UPDATE` and `DELETE` are not allowed on append-optimized tables in a repeatable read or serizalizable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE` and triggers are not supported with append-optimized tables.
```
Comment (soumyadeep2007) [inline gpdb-doc/markdown/admin_guide/ddl/ddl-storage.html.md:47]: Let's not remove this. It is only against Btree indexes that we can cluster AO/CO tables.

File: gpdb-doc/markdown/ref_guide/sql_commands/CLUSTER.html.md [hunk 2]
```diff
@@ -14,7 +14,7 @@ CLUSTER [VERBOSE]
...
+`CLUSTER` orders a table based on an index. Clustering an index means that the records are physically ordered on disk according to the index information. If the records you need are distributed randomly on disk, then the database has to seek across the disk to get the records requested. If those records are stored more closely together, then the fetching from disk is more sequential. A good example for a clustered index is on a date column where the data is ordered sequentially by date. A query against a specific date range will result in an ordered fetch from the disk, which leverages faster sequential access.
```
Comment (soumyadeep2007) [inline gpdb-doc/markdown/ref_guide/sql_commands/CLUSTER.html.md:17]: We should keep:
`CLUSTER` on non-B-tree indexes is not supported on append-optmized storage tables.

File: gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md [hunk 3]
```diff
@@ -498,7 +498,7 @@ vacuum_index_cleanup
 
     For inherited tables, unique constraints, primary key constraints, indexes and table privileges are *not* inherited in the current implementation.
 
--   For append-optimized tables, `UPDATE` and `DELETE` are not allowed in a repeatable read or serializable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE`, and triggers are not supported with append-optimized tables. `CLUSTER` on append-optimized tables is only supported over B-tree indexes.
+-   For append-optimized tables, `UPDATE` and `DELETE` are not allowed in a repeatable read or serializable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE`, and triggers are not supported with append-optimized tables.
```
Comment (soumyadeep2007) [inline gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md:501]: We need to keep:
`CLUSTER` on append-optimized tables is only supported over B-tree indexes.

**Review by soumyadeep2007 — APPROVED** [review]
LGTM! (only change suggested is not to remove the proviso about btree indexes and AO/CO tables)
Out of scope for this PR but nice to add somewhere are the best practices for clustering both AO/CO tables and heap tables in GPDB:
https://greenplum.org/gp7-clustering-aoco-tables/#Best-practices

**Comment by VMWareJennyBeth** [comment]
Great suggestion. I will create a JIRA. Thanks.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On February 15, 2023 at 13:52:31 PST, Soumyadeep Chakraborty ***@***.***> wrote:
!! External Email

@soumyadeep2007 approved this pull request.

LGTM!
Out of scope for this PR but nice to add somewhere are the best practices for clustering both AO/CO tables and heap tables in GPDB:
https://greenplum.org/gp7-clustering-aoco-tables/#Best-practices<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgreenplum.org%2Fgp7-clustering-aoco-tables%2F%23Best-practices&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=bWS2FXMQxGkOkOgO6K%2BpnH5vl9%2BvWY%2FqEqyP%2FgnVL50%3D&reserved=0>

________________________________

In gpdb-doc/markdown/admin_guide/ddl/ddl-storage.html.md<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgreenplum-db%2Fgpdb%2Fpull%2F14983%23discussion_r1107761136&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=Rg2Y6Xuu1%2Fn6mSsN7qR8Bqzd8slQgh7XCliIeqLvXTs%3D&reserved=0>:

> @@ -44,7 +44,7 @@ Use the `WITH` clause of the `CREATE TABLE` command to declare the table storage

 > **Note** You use the `appendoptimized=value` syntax to specify the append-optimized table storage type. `appendoptimized` is a thin alias for the `appendonly` legacy storage option. Greenplum Database stores `appendonly` in the catalog, and displays the same when listing storage options for append-optimized tables.

-`UPDATE` and `DELETE` are not allowed on append-optimized tables in a repeatable read or serizalizable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE` and triggers are not supported with append-optimized tables. `CLUSTER` on append-optimized tables is only supported over B-tree indexes.
+`UPDATE` and `DELETE` are not allowed on append-optimized tables in a repeatable read or serizalizable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE` and triggers are not supported with append-optimized tables.


Let's not remove this. It is only against Btree indexes that we can cluster AO/CO tables.

________________________________

In gpdb-doc/markdown/ref_guide/sql_commands/CLUSTER.html.md<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgreenplum-db%2Fgpdb%2Fpull%2F14983%23discussion_r1107761904&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=SBjJA0yXBzKDJIdwv3HO1wEZn%2BlNDB3M5X5DrtiPNmU%3D&reserved=0>:

> @@ -14,7 +14,7 @@ CLUSTER [VERBOSE]

 ## <a id="section3"></a>Description

-`CLUSTER` orders a heap storage table based on an index. `CLUSTER` on non-B-tree indexes is not supported on append-optmized storage tables. Clustering an index means that the records are physically ordered on disk according to the index information. If the records you need are distributed randomly on disk, then the database has to seek across the disk to get the records requested. If those records are stored more closely together, then the fetching from disk is more sequential. A good example for a clustered index is on a date column where the data is ordered sequentially by date. A query against a specific date range will result in an ordered fetch from the disk, which leverages faster sequential access.
+`CLUSTER` orders a table based on an index. Clustering an index means that the records are physically ordered on disk according to the index information. If the records you need are distributed randomly on disk, then the database has to seek across the disk to get the records requested. If those records are stored more closely together, then the fetching from disk is more sequential. A good example for a clustered index is on a date column where the data is ordered sequentially by date. A query against a specific date range will result in an ordered fetch from the disk, which leverages faster sequential access.


We should keep:
CLUSTER on non-B-tree indexes is not supported on append-optmized storage tables.

________________________________

In gpdb-doc/markdown/ref_guide/sql_commands/CREATE_TABLE.html.md<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgreenplum-db%2Fgpdb%2Fpull%2F14983%23discussion_r1107762335&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=N%2FQnuapdBD1TuZ%2FKBCMPL%2FjNLC3GUiGh3xLyTeSLMdY%3D&reserved=0>:

> @@ -498,7 +498,7 @@ vacuum_index_cleanup

     For inherited tables, unique constraints, primary key constraints, indexes and table privileges are *not* inherited in the current implementation.

--   For append-optimized tables, `UPDATE` and `DELETE` are not allowed in a repeatable read or serializable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE`, and triggers are not supported with append-optimized tables. `CLUSTER` on append-optimized tables is only supported over B-tree indexes.
+-   For append-optimized tables, `UPDATE` and `DELETE` are not allowed in a repeatable read or serializable transaction and will cause the transaction to end prematurely. `DECLARE...FOR UPDATE`, and triggers are not supported with append-optimized tables.


We need to keep:
CLUSTER on append-optimized tables is only supported over B-tree indexes.

—
Reply to this email directly, view it on GitHub<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fgreenplum-db%2Fgpdb%2Fpull%2F14983%23pullrequestreview-1300406392&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=IDURstclOSoJo3ubuWvBykZGyAyzy%2FOP6W4F3iqJaiA%3D&reserved=0>, or unsubscribe<https://nam04.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FASD6RB7L2ZHOW3JM3EHFKB3WXVFZZANCNFSM6AAAAAAU5LPF6M&data=05%7C01%7Cjschaffer%40vmware.com%7C84c74a92ab654f3fcd2908db0f9eeefa%7Cb39138ca3cee4b4aa4d6cd83d9dd62f0%7C0%7C0%7C638120947509294573%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=9%2BPleK1rfPuPpllJjoCaLXoAph6nJM9ZVl4HQ45qUR0%3D&reserved=0>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

!! External Email: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender.

**Review by dyozie — CHANGES_REQUESTED** [review]
The table at https://docs.vmware.com/en/VMware-Tanzu-Greenplum/7/greenplum-database/GUID-ref_guide-feature_summary.html still lists `CLUSTER ON` as unsupported for ALTER TABLE.

Also, the Release Note entry for this should still specify for b-tree indexes only.

**Review by dyozie — APPROVED** [review]

**Pull request merged**